include/binary_graph_stream.h

```diff
@@ -84,31 +85,80 @@ class BinaryGraphStream_MT {
+
+  // call this function to tell stream its okay to keep going
+  // call once per query performed regardless if registered query or on-demand query
+  void post_query_resume() { query_block = false; query_index = -1; }
+
+  /*
+   * Call this function to register a query in advance to avoid contraint on 32 KiB boundary
+   * Only one query may be registered at a time and query index must within a 32 KiB boundary not
+   * yet touched by the MT_StreamReader threads.
+   * To register first query, call before processing any updates from the stream
+   * When registering second (or later) query, call register_query after post_query_resume but 
+   * before calls to get_edge() to ensure the registration and query succeed
+   * @param query_idx   the stream update index directly after which the query will be performed
+   * @return            true if the query is successfully registered and false if not
+  */
+  bool register_query(uint64_t query_idx) {
+    uint64_t byte_index = 12 + query_idx * edge_size;
+    if (byte_index <= stream_off) return false;
+    else query_index = byte_index;
+    return true;
+  }
+
   inline void stream_reset() {stream_off = 12;}
   inline uint32_t nodes() {return num_nodes;}
   inline uint64_t edges() {return num_edges;}
   BinaryGraphStream_MT(const BinaryGraphStream_MT &) = delete;
   BinaryGraphStream_MT & operator=(const BinaryGraphStream_MT &) = delete;
   friend class MT_StreamReader;
-
 private:
   int stream_fd;
-  std::atomic<uint64_t> stream_off;
-  uint32_t num_nodes;     // number of nodes in the graph
-  uint64_t num_edges;     // number of edges in the graph stream
-  uint32_t buf_size;      // how big is the data buffer
-  const uint32_t edge_size = sizeof(uint8_t) + 2 * sizeof(uint32_t); // size of a binary encoded edge
-  uint64_t end_of_file;
+  uint32_t num_nodes;    // number of nodes in the graph
+  uint64_t num_edges;    // number of edges in the graph stream
+  uint32_t buf_size;     // how big is the data buffer
+  uint64_t end_of_file;  // the index of the end of the file
+  std::atomic<uint64_t> stream_off;  // where do threads read from in the stream
+  std::atomic<uint64_t> query_index; // what is the index of the next query in bytes
+  std::atomic<bool> query_block;     // If true block read_data calls and have thr return NXT_QUERY
+  const uint32_t edge_size = sizeof(uint8_t) + 2 * sizeof(uint32_t); // size of binary encoded edge
+
   inline uint32_t read_data(char *buf) {
+    // we are blocking on a query or the stream is done so don't fetch_add or read
+    if (query_block || stream_off >= end_of_file || stream_off >= query_index) return 0;
+
+    // multiple threads may execute this line of code at once. This can cause edge cases
     uint64_t read_off = stream_off.fetch_add(buf_size, std::memory_order_relaxed);
+
+    // we catch these edge cases using the two below checks
+    if (read_off >= query_index) {
+      stream_off = query_index.load();
```

Comment:
This is true. However, an important requirement to perform a query is that all threads have stopped doing updates (therefore no more calls to read_data()). Assuming that the user of the stream is handling this correctly all we need to worry about here is when a bunch of threads call fetch_add at once when stream_off < query_index. 

We can safely set back to query_index because that's where the stream should stop pre-query. No threads should be calling read_data() during the query so there shouldn't be a race on query_index being changed in a call to register_query() or post_query_resume(). It's the responsibility of the thread performing the query to perform the necessary updates to the query_index before unblocking the other threads.

Although, I'm not super happy with this design as it is very reliant on being used correctly. Do you have any ideas for de-facto safer design or a way to force users to interact with the stream in a safe manner?

Reply:
From what you've told me, it sounds like the system is almost workable. We'll need to add documentation/comments/preconditions explaining how to use this function properly (assuming I haven't missed it), as well as a readers-writer lock.

Specifically, a "Read-preferring RW lock."

https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock

There may be standard implementations, but here's one I'm thinking of:

Each thread stores a 'I'm within the critical zone' atomic. Assuming no writers, this atomic can safely be stored in L1 with no contention. Then, when a query occurs, whichever thread is responsible for it (I assume this is a single thread here) iterates through all the relevant atomics, locking them and preventing them from passing. After the 'Query' thread locks all other threads, we know that no other threads can be within the critical zone, regardless of outside implementation.

Let me know what you think.

Reply:
That works for me